runtime/parachains/src/disputes.rs

```diff
@@ -1200,6 +1232,11 @@ impl<T: Config> Pallet<T> {
 			Error::<T>::SingleSidedDispute,
 		);
 
+		// Reject statements with no accompanying backing votes.
+		ensure!(!backers.is_empty(), Error::<T>::MissingBackingVotes,);
```

Comment:
We could require at least backing threshold here.

Reply:
I think we don't need to, the simple check is fine. All honest backers will try to import all the backing votes they know of (and they should know all of them due to chain scraping). In addition not accepting a dispute import is kind of "dangerous", so minimal requirements are in order.

Question is whether we need this restriction at all. With changes @tdimitrov is working on, we will already only import disputes that are confirmed. If a dispute is confirmed, it should be valid. If it is valid, it should contain backing votes. If it for whatever reason does not, then we have a dispute that does not result in slashes, which is then still fine as the dispute was confirmed.

Anyhow, that code does not exist yet, so having this check is probably good, but it should be as lenient as possible (so keep it as `is_empty`) and it would be good to have a comment saying why we care at all.... The only reason I can think of is spam.

Another thing to consider: The node side does some filtering if we have to import lots of disputes. Due to that filtering it could happen that the backing votes get filtered out (would be provided in the next block). I guess that's what you mean by:

```
Ensure block production ensures backing votes are present
```

Reply:
The point of this check is to address a concern from @rphmeier:
> We will also need to ensure that backing votes always get placed on-chain, even when other votes (approval or explicit) are already present. Otherwise, an attacker could avoid being slashed by placing an explicit vote on-chain before their own backing vote is posted by someone else.

This is done by rejecting (blocks with) new dispute statements that don't have any backing votes.

> All honest backers will try to import all the backing votes they know of (and they should know all of them due to chain scraping). 

The thing is, I'm not sure this is the case. 

First, currently, kusama and polkadot don't use prioritized_selection, but a random_selection, correct?
Second, even with prioritized one, I see some code like 
https://github.com/paritytech/polkadot/blob/a59c120ce8d63ae6e0ba0a365d98ea9142a6b770/node/core/provisioner/src/disputes/prioritized_selection/mod.rs#L375-L378

Am I correct that if an attacker places an explicit for and against dispute votes, its backing vote will be ignored?

One might argue, we should fix that instead of adding an additional check, but it's already hard to keep track of what is currently deployed vs what's in master.

On top of that, we have a notion of "inactive" disputes, so if finality lags leading to 1min block times and a dispute concludes, only up to 3 blocks (which can be DDoSed to make 0) will have a chance to include votes until it becomes "inactive" concluded. And the rules for including votes change again.

> I guess that's what you mean by:
> Ensure block production ensures backing votes are present

Here I mean that for honest node's block production, provisioner will ensure backing votes are present no matter what.
Maybe @tdimitrov could double check that :pray: I'm still not sure if that's the case in all cases (e.g. spam).

Reply:
#6274 is a blocker here

Reply:
The line you referenced (if I remember correctly) only avoids importing votes the chain already knows about. So this is fine. 

Indeed with that only slash backers change it is crucial that backing votes will be present on chain, but I don't see how rejecting an import lacking them will help with that: This only means that if it happened, that we would not know about the dispute at all, which does not sound any better.

Instead the provisioner should ensure to import all relevant votes, so our 2/3rd honest nodes will import the backing votes. Given that we have deterministic import these days (the provisioner knows the votes the chain already knows about) and will only import votes the chain does not know about, hence given enough time some validator will record those statements. On top of that, backing votes are the ones which have the best availability on all validators as every validator has them, even without a dispute even started. So we are already pretty good! Those votes are available everywhere and every honest node will try to put them in.

We could relatively easily do even better though and prioritize importing backing votes in the provisioner to have an additional safeguard. @tdimitrov ? If we do that, there should be a clear comment on why we are doing that (currently only backers are slashed).

Reply:
This issue slipped my mind. Sorry for the late response.

> First, currently, kusama and polkadot don't use prioritized_selection, but a random_selection, correct?

Yes, that's correct. `prioritized_selection` is not used anywhere at the moment.

> Am I correct that if an attacker places an explicit for and against dispute votes, its backing vote will be ignored?

It shouldn't. As @eskimor says - this is just for double votes which were already seen onchain. Fresh double votes will be imported. We decided to do it explicitly for slashing purposes.

Reply:
My concern with only #6274 is the following. Imagine an attacker creates a dispute 1 block before its block production slot. It can then place votes (and the dispute concludes) excluding its backing vote. Now, do we want to rely on all different rules including #6274 to place its backing vote(s) after the dispute concludes? My point is, the code for ensuring that is complicated (current version vs master vs "inactive", etc) and if it's easy to ensure that adding this check in block import should never fail on honest nodes, then it adds some extra (redundant) protection.

Reply:
> The line you referenced (if I remember correctly) only avoids importing votes the chain already knows about. So this is fine.
> It shouldn't. As @eskimor says - this is just for double votes which were already seen onchain. Fresh double votes will be imported. We decided to do it explicitly for slashing purposes.

As described above, an attacker can initiate a dispute with backing votes, but then place only explicit double votes (for and against) on chain.

Reply:
Now I am getting where you are coming from. Wow, how such an innocent change "Let's only slash backers" can have so many consequences. Indeed, then the runtime check would prevent those malicious guys from only importing explicit votes, preventing honest nodes from importing those backing votes.

Still - let's fix this in the provisioner: That check above should be adjusted, that we always import backing votes - no matter what. It is only 5 on production networks. I agree that this becomes messy, hence please add comments to any such conditions that they exist because we are only slashing backers.